### PR TITLE
More efficient Vespa indexing

### PR DESCRIPTION
**Pull request by pablonyx**
## Description
Curent process:
- Check for existing docs, delete existing values, index into Vespa

New approach:
- Use`doc_updated_at` as a versioning scheme.
- Instead of checking for existing docs with visit API, we can just a) ensure consisted updated at time, index the documents, and then delete en masse documents with a different "version" (ie. `doc_updated_at` value)

Quick notes:
- We ensure all chunks for a given document share the same doc_updated_at, raise an error if there is a mismatch.  
- Can remove `fresh_index` flag as deletion of old chunks is much less intensive now
- We now use a quick `_does_doc_exist_in_vespa` check to maintain the functionality of determining which documents are wholly new

## How Has This Been Tested?
Various indexing + reindexing + pull updates for 4 connectors (file, web, confluence, Github)
- + Ensure that `already_existed` is accurate

## Backporting (check the box to trigger backport action)
Note: You have to check that the action passes, otherwise resolve the conflicts manually and tag the patches.
- [ ] This PR should be backported (make sure to check that the backport attempt succeeds)

### CODE REVIEW

File: backend/onyx/document_index/vespa/shared_utils/vespa_request_builders.py [hunk 1]
```diff
@@ -106,3 +107,25 @@ def build_vespa_id_based_retrieval_yql(
 
     id_based_retrieval_yql_section += ")"
     return id_based_retrieval_yql_section
+
+
+def build_deletion_selection(doc_id: str, version_cutoff: int, doc_type: str) -> str:
```
Comment (rkuo-danswer) [inline backend/onyx/document_index/vespa/shared_utils/vespa_request_builders.py]: Any way to test this?  Or should we put this in a follow on PR?

File: backend/onyx/document_index/vespa/indexing_utils.py [hunk 2]
```diff
@@ -248,3 +252,84 @@ def clean_chunk_id_copy(
...
+    url = (
+        f"{SEARCH_ENDPOINT}"
+        f'?yql=select+*+from+sources+{index_name}+where+document_id+contains+"{encoded_doc_id}"'
+        "&hits=0"
+    )
+
+    logger.debug(f"Checking existence for doc_id={doc_id} with URL={url}")
+    resp = http_client.get(url)
+
+    if resp.status_code == 200:
```
Comment (rkuo-danswer) [inline backend/onyx/document_index/vespa/indexing_utils.py:279]: The flow here is probably easier to follow if we treat 200 as the happy path and the other paths as exceptions/failures.

File: backend/onyx/document_index/vespa/shared_utils/vespa_request_builders.py [hunk 3]
```diff
@@ -106,3 +107,25 @@ def build_vespa_id_based_retrieval_yql(
...
+    # Escape single quotes by doubling them for Vespa selection expressions
+    escaped_doc_id = doc_id.replace("'", "''")
+
+    return (
+        f"({doc_type}.document_id=='{escaped_doc_id}') and "
+        f"({doc_type}.doc_updated_at < {version_cutoff})"
+    )
+
+
+def build_selection_query_param(filter_str: str) -> str:
```
Comment (rkuo-danswer) [inline backend/onyx/document_index/vespa/shared_utils/vespa_request_builders.py]: Does this get used in more than one place?  Maybe it should just be moved inline with the calling function?

File: backend/onyx/document_index/vespa/index.py [hunk 4]
```diff
@@ -307,47 +314,61 @@ def register_multitenant_indices(
...
+                    )
+                    chunk.source_document.doc_updated_at = chunk_val
+                doc_to_doc_updated_at[doc_id] = chunk_val
+            else:
+                # If the chunk's doc_updated_at is missing, fill it from the stored value
+                if chunk_val is None:
+                    chunk.source_document.doc_updated_at = doc_to_doc_updated_at[doc_id]
+                else:
+                    # Raise if doc_updated_at conflicts with what we've already stored for this doc_id
+                    assert chunk_val == doc_to_doc_updated_at[doc_id], (
```
Comment (rkuo-danswer) [inline backend/onyx/document_index/vespa/index.py]: we probably mean to raise an exception here

File: backend/onyx/document_index/vespa/index.py [hunk 5]
```diff
@@ -307,47 +314,61 @@ def register_multitenant_indices(
...
-        multiple chunk batches calling this function multiple times, otherwise only the last set of
-        chunks will be kept"""
+        """
+        Index a list of chunks into Vespa, optionally removing older chunk versions.
+        Raises an AssertionError if multiple chunks for the same doc_id have conflicting doc_updated_at values.
+        """
+        current_timestamp = int(datetime.utcnow().timestamp())
+
+        # Step 1: Enforce a single doc_updated_at for each doc_id across all chunks.
+        doc_to_doc_updated_at = {}
```
Comment (rkuo-danswer) [inline backend/onyx/document_index/vespa/index.py]: This is an interesting use of doc_updated_at ... I'm unclear on if it was being used for anything important before this that might be affected?  What are the tradeoffs vs using a new field?

File: backend/onyx/document_index/vespa/shared_utils/vespa_request_builders.py (hunk 1 above)
Comment (pablonyx) [reply backend/onyx/document_index/vespa/shared_utils/vespa_request_builders.py]: There certainly is! If you query vespa for a specific number of docs, pull in updates to a connector (which now has fewer chunks), you should see fewer results for the doc ID when you query. Tested this one manually but could probably use a test!

File: backend/onyx/document_index/vespa/indexing_utils.py (hunk 2 above)
Comment (pablonyx) [reply backend/onyx/document_index/vespa/indexing_utils.py:279]: This was definitely the intention- is there a way we can make it clearer that 200 is the happy path in the existing code?

File: backend/onyx/document_index/vespa/shared_utils/vespa_request_builders.py (hunk 1 above)
Comment (pablonyx) [reply backend/onyx/document_index/vespa/shared_utils/vespa_request_builders.py]: There certainly is! If you query vespa for a specific number of docs, pull in updates to a connector (which now has fewer chunks), you should see fewer results for the doc ID when you query. Tested this one manually but could probably use a test!

File: backend/onyx/document_index/vespa/index.py (hunk 4 above)
Comment (pablonyx) [reply backend/onyx/document_index/vespa/index.py]: Good catch

File: backend/onyx/document_index/vespa/indexing_utils.py (hunk 2 above)
Comment (rkuo-danswer) [reply backend/onyx/document_index/vespa/indexing_utils.py:279]: Could we try:

if 404 ... return
if not 200 ... return

200 - happy path

Comment (Weves) [reply backend/onyx/document_index/vespa/indexing_utils.py:279]: +1 to this pattern!

File: backend/onyx/document_index/vespa/index.py [hunk 6]
```diff
@@ -312,17 +312,18 @@ def index(
...
+        Index a list of chunks into Vespa. We rely on 'current_index_time'
         to keep track of when each chunk was added/updated in the index. We also raise a ValueError
-        if any chunk is missing a 'last_indexed_at' timestamp.
+        if any chunk is missing a 'current_index_time' timestamp.
         """
 
         # Clean chunks if needed (remove invalid chars, etc.)
         cleaned_chunks = [clean_chunk_id_copy(chunk) for chunk in chunks]
 
         #  We will store the set of doc_ids that previously existed in Vespa
```
Comment (rkuo-danswer) [inline backend/onyx/document_index/vespa/index.py:323]: nit: extra space

**Review by Weves — COMMENTED** [review]
Approving for now, will need to think more about whether the `current_index_time` should be part of the chunk obj

File: backend/onyx/document_index/vespa/index.py [hunk 7]
```diff
@@ -347,9 +348,9 @@ def index(
                     executor=executor,
                 )
 
-            # c) Remove chunks with using versioning scheme 'last_indexed_at'
+            # c) Remove chunks with using versioning scheme 'current_index_time'
```
Comment (rkuo-danswer) [inline backend/onyx/document_index/vespa/index.py:351]: can deleting old chunks after feeding new chunks cause any issues?  As in returning out of date chunks in results?

File: backend/onyx/document_index/vespa/shared_utils/vespa_request_builders.py (hunk 1 above)
Comment (rkuo-danswer) [reply backend/onyx/document_index/vespa/shared_utils/vespa_request_builders.py]: ok, let's add a linear issue to follow up

File: backend/onyx/document_index/vespa/index.py (hunk 7 above)
Comment (Weves) [reply backend/onyx/document_index/vespa/index.py:351]: Yes I believe so, although the old approach had issues too (e.g. post deletion, pre re-insertion doc would not show up). Hard to say which is preferable.

**Pull request merged**